lib/gollum-lib/wiki.rb

```diff
@@ -504,31 +504,26 @@ def size(ref = nil)
     # query - The string to search for
     #
     # Returns an Array with Objects of page name and count of matches
-    def search(query)
+    def search(query, return_terms = false)
       options = {:path => page_file_dir, :ref => ref}
-      results = {}
-      @repo.git.grep(query, options).each do |hit|
-        name = hit[:name]
-        count = hit[:count]
-        # Remove ext only from known extensions.
-        # test.pdf => test.pdf, test.md => test
-        file_name = Page::valid_page_name?(name) ? name.chomp(::File.extname(name)) : name
-        results[file_name] = count.to_i
-      end
-
-      # Use ls_files '*query*' to search for file names. Grep only searches file content.
-      # Spaces are converted to dashes when saving pages to disk.
-      @repo.git.ls_files(query, options).each do |path|
-        # Remove ext only from known extensions.
-        file_name          = Page::valid_page_name?(path) ? path.chomp(::File.extname(path)) : path
-        # If there's not already a result for file_name then
-        # the value is nil and nil.to_i is 0.
-        results[file_name] = results[file_name].to_i + 1;
-      end
-
-      results.map do |key, val|
-        { :count => val, :name => key }
+      search_terms = query.scan(/"([^"]+)"|(\S+)/).flatten.compact.map {|term| Regexp.escape(term)}
+      search_terms_regex = search_terms.join('|')
+      query = /^(.*(?:#{search_terms_regex}).*)$/i
+      results = @repo.git.grep(search_terms, options) do |name, data|
+        result = {:count => 0}
+        result[:name] = extract_page_file_dir(name)
+        result[:filename_count] = result[:name].scan(/#{search_terms_regex}/i).size
+        found_in_blob = []
+        if data
+          data.scan(query) do |match|
+            found_in_blob << match.first
+            result[:count] += match.first.scan(/#{search_terms_regex}/i).size
+          end
+        end
+        result[:context] = found_in_blob
+        ((result[:count] + result[:filename_count]) == 0) ? nil : result
```

Comment:
What is this line supposed to do? It is not a return statement but it's also not setting any variable it seems.

Reply:
It is supposed to function as a return statement, but not sure if you can use explicit `return` in a block. Either way the idea is to only store results for files that actually contain a hit (either in-text or filename).